packages/contract-utils/upgradeable/src/storage.rs

```diff
@@ -0,0 +1,106 @@
+use soroban_sdk::{contracttype, panic_with_error, symbol_short, Env, Symbol};
+
+use crate::upgradeable::UpgradeableError;
+
+pub const UPGRADE_KEY: Symbol = symbol_short!("UPGRADE");
+
+/// Represents the current upgrade state of the contract. Used to determine if
+/// migration or rollback operations are allowed.
+#[contracttype]
+pub enum UpgradeState {
```

Comment:
I really like this approach of keeping an enum to track states. If, in the future, we want to further optimize this cost, there is a way to enforce the state using types in compile time, have a look at: https://docs.rs/state-shift/latest/state_shift/

It also hides the underlying generics from the end-developer. However, since stellar/soroban does not support generics at some places, I'm not sure whether this will work or not.

May worth trying though at a later stage, not a blocker for this PR